docs/src/modules/components/MarkdownDocs.js

```diff
@@ -55,6 +55,32 @@ export default function MarkdownDocs(props) {
   const Provider = isJoy ? CssVarsProvider : React.Fragment;
   const Wrapper = isJoy ? BrandingProvider : React.Fragment;
 
+  const directTsDemos = rendered
+    .map((renderedMarkdownOrDemo) => {
+      if (typeof renderedMarkdownOrDemo === 'string') {
+        return null;
+      }
+
+      const { demo = '', hideToolbar } = renderedMarkdownOrDemo;
+
+      if (demo.endsWith('.tsx') || demo.endsWith('.ts')) {
+        if (!hideToolbar) {
+          return demo;
+        }
+      }
+
+      return null;
+    })
+    .filter((v) => v);
+
+  if (directTsDemos.length > 0) {
+    throw new Error([
+      `The following demos use TS directly:\n`,
+      directTsDemos.join('\n'),
+      `\nPlease run "yarn docs:typescript:formatted" to generate a JS version and reference that instead, or hide the toolbar like this {{"demo": "YourDemo.tsx", "hideToolbar": true, …}}`,
+    ]);
```

Comment:
For the error, I would propose adding the check later, only once we start to the the data:

```diff
diff --git a/docs/src/modules/components/Demo.js b/docs/src/modules/components/Demo.js
index db5d3915c7..1f4f51583b 100644
--- a/docs/src/modules/components/Demo.js
+++ b/docs/src/modules/components/Demo.js
@@ -328,6 +328,20 @@ export default function Demo(props) {
         ].join('\n'),
       );
     }
+
+    if (!demoOptions.demo.endsWith('.js') && demoOptions.hideToolbar !== true) {
+      throw new Error(
+        [
+          `The following demos use TS directly: ${demoOptions.demo}.`,
+          '',
+          'Please run "yarn docs:typescript:formatted" to generate a JS version and reference it:',
+          `{{"demo": "${demoOptions.demo.replace(/\.(.*)$/, '.js')}", …}}.`,
+          '',
+          "Otherwise, if it's not a code demo hide the toolbar:",
+          `{{"demo": "${demoOptions.demo}", "hideToolbar": true, …}}.`,
+        ].join('\n'),
+      );
+    }
   }

   const t = useTranslate();
```

<img width="992" alt="Screenshot 2023-01-29 at 22 46 30" src="https://user-images.githubusercontent.com/3165635/215357410-5c9ce207-316e-45d2-91f7-5a7ad33ced1f.png">